File: src/main/kotlin/acc/business/Facade.java
package acc.business;

import java.util.List;
import acc.business.balance.BalanceItem;
import acc.business.balance.Balance;
import acc.integration.AccountDAO;
import acc.integration.TransactionDAO;
import acc.integration.impl.DocumentDAO;
import acc.model.AnalAcc;
import acc.model.AccGroup;
import acc.model.AccId;
import acc.model.Document;
import acc.model.DocumentId;
import acc.model.DocumentType;
import acc.model.Transaction;
import acc.model.TransactionId;
import acc.util.AccException;
import java.time.LocalDate;
import java.time.Month;
import java.util.Optional;

public class Facade {

    public static Facade instance = new Facade();

    public static final String POCATECNI_STAV = "pocatecni stav";

    public void createAccount(AccGroup skupina, String no, String name) throws AccException {
        AccountDAO.instance.create(skupina, no, name);
    }

    public List<AnalAcc> getAllAccounts() throws AccException {
        List<AnalAcc> result = AccountDAO.instance.getAll();
        return result;
    }

    public AnalAcc pur() throws AccException {
        return AccountDAO.instance.getPocatecniUcetRozvazny();
    }

    public void deleteAccount(AccId id) throws AccException {
        AccountDAO.instance.delete(id);
    }

    public List<AnalAcc> getBalanceAccounts() throws AccException {
        return AccountDAO.instance.getBalancing();
    }

    public Optional<AnalAcc> getAccountByNumber(String number) throws AccException {
        return AccountDAO.instance.getByNumber(number);
    }

    public List<AnalAcc> getAccountsByGroup(AccGroup accg) throws AccException {
        return AccountDAO.instance.getByGroup(accg);
    }

    public List<AnalAcc> getAccountsByClass(AccGroup accg) throws AccException {
        return AccountDAO.instance.getByClass(accg);
    }

    public void createTransaction(long amount, AnalAcc madati, AnalAcc dal, Optional<Document> document, Optional<Document> bindingDocument) throws AccException {
        TransactionDAO.instance.create(amount, madati,
                dal, document, bindingDocument);
    }

    public List<Transaction> getAllTransactions() throws AccException {
        return TransactionDAO.instance.getAll();
    }

    public void updateTransaction(TransactionId id, long amount,
            AnalAcc madati, AnalAcc dal, Optional<Document> document,
            Optional<Document> bindingDocument) throws AccException {
        TransactionDAO.instance.update(id, amount, madati, dal,
                document, bindingDocument);
    }

    public void deleteTransaction(TransactionId id) throws AccException {
        TransactionDAO.instance.delete(id);
    }

    public List<Transaction> getTransactions(Optional<LocalDate> optOd, Optional<LocalDate> optDo,
            Optional<AnalAcc> acc, Optional<Document> optDocument, Optional<Document> optRelatedDocument) throws AccException {
        return TransactionDAO.instance.get(optOd, optDo, acc, optDocument, optRelatedDocument);
    }

    public List<BalanceItem> createBalance(Month month) throws AccException {
        return Balance.instance.createBalance(month);
    }

    public void createInit(long amount, AnalAcc acc, MadatiDal madatiDal) throws AccException {
        switch (madatiDal) {
            case MA_DATI: {
                TransactionDAO.instance
                        .create(amount, acc, pur(), Optional.empty(), Optional.empty());
                break;
            }
            case DAL: {
                TransactionDAO.instance
                        .create(amount, pur(), acc, Optional.empty(), Optional.empty());
            }
        }
    }

    public List<Transaction> getInits(Optional<AnalAcc> acc) throws AccException {
        return TransactionDAO.instance.getInits(acc);
    }

    // Document ***************************************************************
    public void createDocument(DocumentType type, String name,
            LocalDate date, String description) {
        DocumentDAO.instance.create(type, name, date, description);
    }

    public void updateDocument(DocumentId id, DocumentType type, String name,
            LocalDate date, String description) {
        DocumentDAO.instance.update(id, type, name, date, 
                description);
    }

    public void deleteDocument(DocumentId id) throws AccException {
        DocumentDAO.instance.delete(id);
    }

    public List<Document> getAllDocuments() throws AccException {
        return DocumentDAO.instance.getAll();
    }

    public List<Document> getDocumentByName(String name) throws AccException {
        return DocumentDAO.instance.getByRegexp(name);
    }

}

File: src/main/kotlin/acc/business/Facade.kt
package acc.business

import acc.integration.AccountCache
import acc.integration.DocumentDAO
import acc.integration.TransDAO
import acc.model.*
import acc.util.AccException
import java.time.LocalDate

object Facade {

    // Ucty
    val allAccounts: List<AnalAcc>
        @Throws(AccException::class)
        get() = AccountCache.allAccs

    val balanceAccounts: List<AnalAcc>
        @Throws(AccException::class)
        get() = AccountCache.balanceAccs

    val incomeAccounts: List<AnalAcc>
        @Throws(AccException::class)
        get() = AccountCache.incomeAccs

    val dodavatele: List<AnalAcc>  // 321
        get() = AccountCache.dodavatele

    val pokladny: List<AnalAcc>
        get() = AccountCache.pokladny


    @Throws(AccException::class)
    fun createAccount(group: AccGroup, no: String, name: String, initAmount: Long) {
        AccountCache.createAcc(group, no, name, initAmount)
    }

    @Throws(AccException::class)
    fun updateAccount(acc: AnalAcc, name: String, initAmount: Long) {
        AccountCache.updateAcc(acc, name, initAmount)
    }

    @Throws(AccException::class)
    fun deleteAccount(acc: AnalAcc) {
        AccountCache.deleteAcc(acc)
    }

    @Throws(AccException::class)
    fun accountIsUsed(acc: AnalAcc?): Boolean {
        return transactionsByFilter(TransactionFilter(acc = acc)).isNotEmpty()
    }

    // Transakce ***************************************************************
    val allTransactions: List<Transaction>
        @Throws(AccException::class)
        get() = TransDAO.allTrans


    @Throws(AccException::class)
    fun createTransaction(amount: Long, madati: AnalAcc, dal: AnalAcc,
                          document: Document, relatedDocument: Document?) {
        TransDAO.createTrans(null, amount, madati,
                dal, document, relatedDocument)
    }


    @Throws(AccException::class)
    fun transactionsByFilter(tf: TransactionFilter?): List<Transaction> {
        return TransDAO.transByFilter(tf)
    }

    @Throws(AccException::class)
    fun updateTransaction(id: TransactionId, amount: Long,
                          madati: AnalAcc, dal: AnalAcc, document: Document,
                          relatedDocument: Document?) {
        TransDAO.updateTrans(id, amount, madati, dal,
                document, relatedDocument)
    }

    @Throws(AccException::class)
    fun deleteTransaction(id: TransactionId) {
        TransDAO.deleteTrans(id)
    }

    // Doklady  ***************************************************************
    val allDocuments: List<Document>
        @Throws(AccException::class)
        get() = DocumentDAO.allDocs

    private val allInvoices: List<Document>
        @Throws(AccException::class)
        get() = DocumentDAO.allDocs
                .filter { it.type == DocType.INVOICE }

    val unpaidInvoices: List<Document>
        @Throws(AccException::class)
        get() {
            val paidInvoices = allTransactions.mapNotNull { it.relatedDoc }
            val ui = allInvoices as MutableList
            ui.removeAll(paidInvoices)
            return ui
        }

    @Throws(AccException::class)
    fun documentsByFilter(docFilter: DocFilter?): List<Document> {
        return if (docFilter == UnpaidInvoicesFilter)
            unpaidInvoices
        else
            DocumentDAO.docsByFilter(docFilter) //as MutableList<Document>
    }

    @Throws(AccException::class)
    fun createDocument(_type: DocType, _number: Int, _date: LocalDate, _description: String): Document {
        return Document(
                id = DocumentDAO.createDoc(_type, _number, _date, _description),
                type = _type,
                number =_number,
                date = _date,
                description = _description);
    }

    @Throws(AccException::class)
    fun updateDocument(doc: Document,
                       date: LocalDate, description: String) {
        DocumentDAO.updateDoc(doc, date, description)
    }

    @Throws(AccException::class)
    fun deleteDocument(doc: Document) {
        DocumentDAO.deleteDoc(doc)
    }

    @Throws(AccException::class)
    fun genDocumentNumber(docType: DocType): Int {
        return DocumentDAO.findFreeDocNumber(docType)
    }

}